src/main/kotlin/org/jetbrains/research/ml/coding/assistant/solutionSpace/builder/SolutionSpaceGraphBuilder.kt

```diff
@@ -0,0 +1,90 @@
+package org.jetbrains.research.ml.coding.assistant.solutionSpace.builder
+
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.SolutionSpace
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.removeVertexList
+import org.jetbrains.research.ml.coding.assistant.solutionSpace.utils.replaceVertex
+import org.jetbrains.research.ml.coding.assistant.unification.model.DynamicIntermediateSolution
+import org.jgrapht.alg.cycle.SzwarcfiterLauerSimpleCycles
+import org.jgrapht.graph.SimpleDirectedWeightedGraph
+
+typealias DynamicIntermediateSolutionVertexChain = List<SolutionSpaceGraphVertex>
+internal typealias SolutionSpaceGraph = SimpleDirectedWeightedGraph<SolutionSpaceGraphVertex, SolutionSpaceGraphEdge>
+
+class SolutionSpaceGraphBuilder {
+    internal val cache = TreeContextCache()
+    internal val graph = SolutionSpaceGraph(SolutionSpaceGraphEdgeFactory(cache))
+
+    fun addDynamicSolution(dynamic: DynamicIntermediateSolution) {
+        val chain = dynamic.map { SolutionSpaceGraphVertex(it) }
+        val dynamicVertices = graph.addDynamicChainOrMerge(chain)
+        for ((source, target) in dynamicVertices.windowed(2)) {
+            val sourceVertex = graph.fetchVertex(source)
+            val targetVertex = graph.fetchVertex(target)
+            if (sourceVertex == targetVertex) {
+                continue
+            }
+            val edge = if (graph.containsEdge(sourceVertex, targetVertex)) {
+                graph.getEdge(sourceVertex, targetVertex)
+            } else {
+                graph.addEdge(sourceVertex, targetVertex)
+            }
+            graph.setEdgeWeight(edge, edge.calculatedWeight)
+        }
+    }
+
+    private fun clearCycles() {
+        val simpleCyclesDetector = SzwarcfiterLauerSimpleCycles(graph)
+        for (simpleCycle in simpleCyclesDetector.findSimpleCycles()) {
+            graph.removeVertexList(simpleCycle)
+        }
+        graph.vertexSet()
+            .filter {
+                graph.edgesOf(it).isEmpty() && !it.representativeSolution.metaInfo.isFinalSolution
+            }
+            .forEach(graph::removeVertex)
+    }
+
+    private fun clearPsiFiles() {
```

Comment:
clear -> delete/remove?

Reply:
And another question. Maybe should we delete the folder? and nested files will be deleted automatically?